source/code/source/playfab/PlayFabEventPipeline.cpp

```diff
@@ -286,9 +298,17 @@ namespace PlayFab
     {
         try
         {
-            // batch was successfully sent out, find it in the batch tracking map using customData pointer as a key
-            const auto foundBatchIterator = this->batchesInFlight.find(customData);
-            if (foundBatchIterator == this->batchesInFlight.end())
+            bool batchMissingFromFlight = false;
+            std::unordered_map<void*, std::vector<std::shared_ptr<const IPlayFabEmitEventRequest>>>::iterator foundBatchIterator;
+
+            { // LOCK batchesInFlight mutex
+                std::unique_lock<std::mutex> lock(inFlightMutex);
+                // batch was successfully sent out, find it in the batch tracking map using customData pointer as a key
+                foundBatchIterator = this->batchesInFlight.find(customData);
+                batchMissingFromFlight = foundBatchIterator == this->batchesInFlight.end();
+            } // UNLOCK batchesInFlight
+            
+            if (batchMissingFromFlight)
```

Comment:
instead of checking for the batch under one lock and then removing that batch under another, why not just pop the batch out in one mutually exclusive operation?

```
std::vector<std::shared_ptr<const IPlayFabEmitEventRequest>> batchWritten;
{ // LOCK
    std::unique_lock lock(..);
    auto iter = this->batchesInFlight.find(customData);
    if (iter == this->batchesInFlight.end())
    {
        // this is a bug.
        LOG_PIPELINE(...);
        return;
    }

    batchWritten = std::move(iter->second);
    this->batchesInFlight.erase(iter);
} // UNLOCK

auto requestBatchPtr = ...;
```

this has the benefit of taking one lock instead of two and reducing some if-else nesting (a safety AND readability win)

it's usually pretty unsafe (or at least an anti pattern) in multithreaded code to access some shared stated in a mutually exclusive context, keep a record of it, and then use that record later to modify that shared state.

consider the following contrived example + pseudocode
```
int intRecord;
// take mutual exclusive access to the shared integer state
{
    intMutex.lock();
    intRecord = sharedInt.load();
}

// do something outside of the lock

// re-take mutual exclusive access to the shared integer state
if (intRecord < INT32_MAX)
{
    intMutex.lock();
    sharedInt.store(intRecord + 1); // unsafe. another thread could have called into this same function, performed this increment, and now we are overwriting their value.
    // another unsafe idea, sharedInt = sharedInt + 1; 
}
```